ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java

```diff
@@ -78,7 +77,6 @@
 import io.confluent.ksql.util.KsqlConfig;
 import io.confluent.ksql.util.KsqlException;
 import io.confluent.ksql.util.VisitorUtil;
-import java.util.ArrayList;
```

Comment:
rather than exposing `TypeContext` in `getExpressionSqlType` I think it would be better to just add a parameter that's a map from String to SqlType that contains the types for all the lambda parameters. `TypeContext` really feels like it should be an internal detail of this class. I know we have it being passed around in `SqlToJavaVisitor` and `CodeGenRunner`, but I don't think we should use a common type unless there's really a common set of information that makes sense to pass around. In `TypeContext` case there's only 2 fields (well 3, but I think we should drop the list and make the map immutable as described below), and 1 (`sqlType`) is only used from `ExpressionTypeManager`.

Reply:
I'm not sure I follow. Would we add the map parameter to all the methods in `ExpressionTypeManager`? Or change `TypeContext` back into `ExpressionTypeManagerContext` and include in that a mapping from string to SqlType for lambdas? We do still need the lambda variable mapping passed around, we can pass _in_ a list from `codeGenRunner` and `SqlToJavaVisitor` and then store it back in the `ExpressionTypeManagerContext` to be passed down where the lambda literal is resolved. I'm not sure I really prefer that method though since we'd also need to change `SqlToJavaVisitor` and potentially `CodeGenRunner` to pass around a map of lambda types since they also need the information shared between method calls. We could go back to an earlier iteration and have a separate context for each of the three classes, that would keep `SqlType` just in `ExpressionTypeManager`, but we'd be doing a lot of list copying from one context type to a different context type

Reply:
@rodesai what are your thoughts on creating a Context class that `SqlToJavaVisitor` and `CodeGenRunner` both use and then a `TypeContext` with `sqlType` field that ExpressionTypeManager uses?

Reply:
> change TypeContext back into ExpressionTypeManagerContext and include in that a mapping from string to SqlType for lambdas?
This would be better IMO.

> We could go back to an earlier iteration and have a separate context for each of the three classes, that would keep SqlType just in ExpressionTypeManager, but we'd be doing a lot of list copying from one context type to a different context type

Personally I prefer keeping a separate context type internal to each class that contains a visitor because it keeps things decoupled. Yes, you might end up with a few context types that look similar but I think this is generally more extensible and easier to understand. Over time, we may want to do different things from different visitors that require different state to be passed around. If we use a common context type, we'll have to have it support the union of the use cases from the different visitors (we already see this today with the `sqlType` method which only `ExpressionTypeManager` uses). This makes it harder to understand/maintain the codebase. If I want to change this context I have to think about all the different things that are going to use it. If it's private within some class I know it's only being used there.

> but we'd be doing a lot of list copying from one context type to a different context type
Unless we're on a performance path I'd optimize for readability/maintainability.